File: kotlin/src/main/kotlin/ai/muonium/muongit/Fetch.kt
// Fetch.kt - Fetch, push, and clone operations
// Parity: libgit2 src/libgit2/fetch.c, push.c, clone.c

package ai.muonium.muongit

import java.io.ByteArrayOutputStream
import java.io.File

// --- Fetch ---

/** Result of computing fetch wants. */
data class FetchNegotiation(
    val wants: List<OID>,
    val haves: List<OID>,
    val matchedRefs: List<MatchedRef>,
)

/** A remote ref matched against a fetch refspec. */
data class MatchedRef(
    val remoteName: String,
    val localName: String,
    val oid: OID,
)

/** Match a ref name against a refspec pattern (supports trailing glob). */
internal fun refspecMatch(name: String, pattern: String): String? {
    if (pattern.endsWith("*")) {
        val prefix = pattern.dropLast(1)
        return if (name.startsWith(prefix)) name.removePrefix(prefix) else null
    }
    return if (name == pattern) "" else null
}

/** Apply a refspec to map a remote ref name to a local ref name. */
fun applyRefspec(remoteName: String, refspec: String): String? {
    val (_, src, dst) = parseRefspec(refspec) ?: return null
    val matched = refspecMatch(remoteName, src) ?: return null

    return if (dst.endsWith("*")) {
        "${dst.dropLast(1)}$matched"
    } else {
        dst
    }
}

/** Compute which objects we need to fetch from the remote. */
fun computeFetchWants(
    remoteRefs: List<RemoteRef>,
    refspecs: List<String>,
    gitDir: File,
): FetchNegotiation {
    val wants = mutableListOf<OID>()
    val matchedRefs = mutableListOf<MatchedRef>()
    val seen = mutableSetOf<OID>()

    for (rref in remoteRefs) {
        for (refspec in refspecs) {
            val localName = applyRefspec(rref.name, refspec) ?: continue
            matchedRefs.add(
                MatchedRef(
                    remoteName = rref.name,
                    localName = localName,
                    oid = rref.oid,
                )
            )

            val alreadyHave = runCatching { resolveReference(gitDir, localName) }
                .getOrNull() == rref.oid

            if (!alreadyHave && seen.add(rref.oid)) {
                wants.add(rref.oid)
            }
        }
    }

    return FetchNegotiation(
        wants = wants,
        haves = collectLocalRefs(gitDir),
        matchedRefs = matchedRefs,
    )
}

/** Collect all local ref OIDs for negotiation. */
internal fun collectLocalRefs(gitDir: File): List<OID> {
    val oids = mutableListOf<OID>()
    for (dir in listOf("refs/heads", "refs/remotes")) {
        val refDir = File(gitDir, dir)
        if (refDir.isDirectory) {
            collectRefsRecursive(refDir, oids)
        }
    }
    return oids
}

private fun collectRefsRecursive(dir: File, oids: MutableList<OID>) {
    val files = dir.listFiles() ?: return
    for (file in files) {
        if (file.isDirectory) {
            collectRefsRecursive(file, oids)
        } else {
            val hex = file.readText().trim()
            if (hex.length == 40) {
                runCatching { OID(hex) }.getOrNull()?.let(oids::add)
            }
        }
    }
}

/** Update local refs after a successful fetch. */
fun updateRefsFromFetch(gitDir: File, matchedRefs: List<MatchedRef>): Int {
    var updated = 0
    for (mref in matchedRefs) {
        writeReference(gitDir, mref.localName, mref.oid)
        updated++
    }
    return updated
}

// --- Push ---

/** A ref update for push. */
data class PushUpdate(
    val srcRef: String,
    val dstRef: String,
    val srcOid: OID,
    val dstOid: OID,
    val force: Boolean,
)

/** Compute push updates. */
fun computePushUpdates(
    pushRefspecs: List<String>,
    gitDir: File,
    remoteRefs: List<RemoteRef>,
): List<PushUpdate> {
    val updates = mutableListOf<PushUpdate>()

    for (refspec in pushRefspecs) {
        val (force, src, dst) = parseRefspec(refspec)
            ?: throw MuonGitException.InvalidObject("invalid push refspec: $refspec")

        val srcOid = resolveReference(gitDir, src)
        val dstOid = remoteRefs.firstOrNull { it.name == dst }?.oid ?: OID.ZERO

        updates.add(
            PushUpdate(
                srcRef = src,
                dstRef = dst,
                srcOid = srcOid,
                dstOid = dstOid,
                force = force,
            )
        )
    }

    return updates
}

/** Build a push report string. */
fun buildPushReport(updates: List<PushUpdate>): String {
    return updates.joinToString("") { update ->
        "${update.dstOid.hex} ${update.srcOid.hex} ${update.dstRef}\n"
    }
}

data class FetchOptions(
    val refspecs: List<String>? = null,
    val transport: TransportOptions = TransportOptions(),
)

data class FetchResult(
    val advertisedRefs: List<RemoteRef>,
    val capabilities: ServerCapabilities,
    val matchedRefs: List<MatchedRef>,
    val updatedRefs: Int,
    val indexedPack: IndexedPack?,
)

data class PushOptions(
    val refspecs: List<String>? = null,
    val transport: TransportOptions = TransportOptions(),
)

data class PushResult(
    val advertisedRefs: List<RemoteRef>,
    val updatedTrackingRefs: Int,
    val report: String,
)

// --- Clone ---

/** Options for clone. */
data class CloneOptions(
    val remoteName: String = "origin",
    val branch: String? = null,
    val bare: Boolean = false,
    val transport: TransportOptions = TransportOptions(),
)

/** Set up a new repository for clone. */
fun cloneSetup(path: String, url: String, options: CloneOptions = CloneOptions()): Repository {
    val repo = Repository.init(path, options.bare)
    addRemote(repo.gitDir, options.remoteName, url)

    if (options.branch != null) {
        val target = "refs/heads/${options.branch}"
        writeSymbolicReference(repo.gitDir, "HEAD", target)
    }

    return repo
}

/** After fetching, set up HEAD and the default branch for a clone. */
fun cloneFinish(gitDir: File, remoteName: String, defaultBranch: String, headOid: OID) {
    val localBranch = "refs/heads/$defaultBranch"
    val remoteRef = "refs/remotes/$remoteName/$defaultBranch"

    writeReference(gitDir, localBranch, headOid)
    writeReference(gitDir, remoteRef, headOid)
    writeSymbolicReference(gitDir, "HEAD", localBranch)
}

/** Extract the default branch from server capabilities. */
fun defaultBranchFromCaps(caps: ServerCapabilities): String? {
    val symref = caps.get("symref") ?: return null
    val colonIdx = symref.indexOf(':')
    if (colonIdx < 0) return null
    val headPart = symref.substring(0, colonIdx)
    val target = symref.substring(colonIdx + 1)
    if (headPart != "HEAD") return null
    return if (target.startsWith("refs/heads/")) {
        target.removePrefix("refs/heads/")
    } else {
        null
    }
}

fun cloneRepository(
    url: String,
    path: String,
    options: CloneOptions = CloneOptions(),
): Repository {
    val repo = cloneSetup(path, url, options)
    val fetch = fetchRemote(
        repository = repo,
        remoteName = options.remoteName,
        options = FetchOptions(transport = options.transport),
    )
    val (branch, headOid) = resolveCloneHead(fetch, options.branch)
    cloneFinish(repo.gitDir, options.remoteName, branch, headOid)
    if (!options.bare) {
        reset(repo.gitDir, repo.workdir, "HEAD", ResetMode.HARD)
    }
    return repo
}

fun fetchRemote(
    repository: Repository,
    remoteName: String,
    options: FetchOptions = FetchOptions(),
): FetchResult {
    val remote = getRemote(repository.gitDir, remoteName)
    val advertisement = advertiseUploadPack(remote.url, options.transport)
    val refspecs = options.refspecs ?: remote.fetchRefspecs
    val negotiation = computeFetchWants(advertisement.refs, refspecs, repository.gitDir)

    if (negotiation.wants.isEmpty()) {
        val updated = updateRefsFromFetch(repository.gitDir, negotiation.matchedRefs)
        return FetchResult(
            advertisedRefs = advertisement.refs,
            capabilities = advertisement.capabilities,
            matchedRefs = negotiation.matchedRefs,
            updatedRefs = updated,
            indexedPack = null,
        )
    }

    val request = buildWantHave(
        wants = negotiation.wants,
        haves = negotiation.haves,
        caps = fetchCapabilities(advertisement.capabilities),
    )
    val response = uploadPack(remote.url, request, options.transport)
    val indexedPack = extractPackFromFetchResponse(response)
        ?.let { indexPackToODB(repository.gitDir, it) }
    val updated = updateRefsFromFetch(repository.gitDir, negotiation.matchedRefs)

    return FetchResult(
        advertisedRefs = advertisement.refs,
        capabilities = advertisement.capabilities,
        matchedRefs = negotiation.matchedRefs,
        updatedRefs = updated,
        indexedPack = indexedPack,
    )
}

fun pushRemote(
    repository: Repository,
    remoteName: String,
    options: PushOptions = PushOptions(),
): PushResult {
    val remote = getRemote(repository.gitDir, remoteName)
    val advertisement = advertiseReceivePack(remote.url, options.transport)
    val refspecs = options.refspecs ?: defaultPushRefspecs(repository.gitDir)
    val updates = computePushUpdates(refspecs, repository.gitDir, advertisement.refs)

    for (update in updates) {
        if (!update.force && !isFastForward(repository.gitDir, update.dstOid, update.srcOid)) {
            throw MuonGitException.NotFastForward()
        }
    }

    val pack = buildPackFromOIDs(
        gitDir = repository.gitDir,
        roots = updates.map { it.srcOid },
        exclude = advertisement.refs.map { it.oid },
    )
    val request = buildPushRequest(updates, pack, advertisement.capabilities)
    val response = receivePack(remote.url, request, options.transport)
    val report = parsePushResponse(response)
    val updatedTrackingRefs = updateTrackingRefsAfterPush(repository.gitDir, remoteName, updates)

    return PushResult(
        advertisedRefs = advertisement.refs,
        updatedTrackingRefs = updatedTrackingRefs,
        report = report,
    )
}

fun Repository.fetch(
    remoteName: String,
    options: FetchOptions = FetchOptions(),
): FetchResult = fetchRemote(this, remoteName, options)

fun Repository.push(
    remoteName: String,
    options: PushOptions = PushOptions(),
): PushResult = pushRemote(this, remoteName, options)

private fun fetchCapabilities(caps: ServerCapabilities): List<String> {
    val requested = mutableListOf<String>()
    if (caps.has("side-band-64k")) {
        requested.add("side-band-64k")
    } else if (caps.has("side-band")) {
        requested.add("side-band")
    }
    if (caps.has("ofs-delta")) {
        requested.add("ofs-delta")
    }
    if (caps.has("include-tag")) {
        requested.add("include-tag")
    }
    return requested
}

private fun extractPackFromFetchResponse(response: ByteArray): ByteArray? {
    val (lines, consumed) = pktLineDecode(response)
    val pack = ByteArrayOutputStream()

    for (line in lines) {
        if (line !is PktLine.Data) {
            continue
        }
        if (startsWithBytes(line.bytes, "ACK ".toByteArray()) || line.bytes.contentEquals("NAK\n".toByteArray())) {
            continue
        }
        when (line.bytes.firstOrNull()?.toInt()) {
            1 -> pack.write(line.bytes, 1, line.bytes.size - 1)
            2 -> Unit
            3 -> throw MuonGitException.InvalidObject(
                String(line.bytes.copyOfRange(1, line.bytes.size), Charsets.UTF_8).trim()
            )
            else -> Unit
        }
    }

    if (pack.size() > 0) {
        return pack.toByteArray()
    }

    if (consumed < response.size) {
        val trailing = response.copyOfRange(consumed, response.size)
        if (startsWithBytes(trailing, "PACK".toByteArray())) {
            return trailing
        }
    }

    return null
}

private fun buildPushRequest(
    updates: List<PushUpdate>,
    pack: ByteArray,
    capabilities: ServerCapabilities,
): ByteArray {
    val requestedCaps = mutableListOf("report-status")
    if (capabilities.has("ofs-delta")) {
        requestedCaps.add("ofs-delta")
    }

    val out = ByteArrayOutputStream()
    for ((index, update) in updates.withIndex()) {
        val line = if (index == 0) {
            "${update.dstOid.hex} ${update.srcOid.hex} ${update.dstRef}\u0000${requestedCaps.joinToString(" ")}\n"
        } else {
            "${update.dstOid.hex} ${update.srcOid.hex} ${update.dstRef}\n"
        }
        out.write(pktLineEncode(line.toByteArray(Charsets.UTF_8)))
    }
    out.write(pktLineFlush())
    out.write(pack)
    return out.toByteArray()
}

private fun parsePushResponse(response: ByteArray): String {
    val (lines, consumed) = pktLineDecode(response)
    val text = StringBuilder()

    for (line in lines) {
        if (line !is PktLine.Data) {
            continue
        }

        val payload = when (line.bytes.firstOrNull()?.toInt()) {
            1, 2 -> line.bytes.copyOfRange(1, line.bytes.size)
            3 -> throw MuonGitException.InvalidObject(
                String(line.bytes.copyOfRange(1, line.bytes.size), Charsets.UTF_8).trim()
            )
            else -> line.bytes
        }
        text.append(String(payload, Charsets.UTF_8))
    }

    if (consumed < response.size) {
        text.append(String(response.copyOfRange(consumed, response.size), Charsets.UTF_8))
    }

    for (line in text.lines()) {
        if (line.startsWith("unpack ") && line != "unpack ok") {
            throw MuonGitException.InvalidObject(line)
        }
        if (line.startsWith("ng ")) {
            throw MuonGitException.InvalidObject(line.removePrefix("ng "))
        }
    }

    return text.toString()
}

private fun resolveCloneHead(fetch: FetchResult, branch: String?): Pair<String, OID> {
    if (branch != null) {
        val refName = "refs/heads/$branch"
        val oid = fetch.advertisedRefs.firstOrNull { it.name == refName }?.oid
            ?: throw MuonGitException.NotFound("remote branch '$branch' not found")
        return Pair(branch, oid)
    }

    defaultBranchFromCaps(fetch.capabilities)?.let { defaultBranch ->
        val refName = "refs/heads/$defaultBranch"
        fetch.advertisedRefs.firstOrNull { it.name == refName }?.oid?.let { oid ->
            return Pair(defaultBranch, oid)
        }
    }

    val headOid = fetch.advertisedRefs.firstOrNull { it.name == "HEAD" }?.oid
    if (headOid != null) {
        fetch.advertisedRefs.firstOrNull {
            it.name.startsWith("refs/heads/") && it.oid == headOid
        }?.let { branchRef ->
            return Pair(branchRef.name.removePrefix("refs/heads/"), headOid)
        }
    }

    for (candidate in listOf("main", "master")) {
        val refName = "refs/heads/$candidate"
        fetch.advertisedRefs.firstOrNull { it.name == refName }?.oid?.let { oid ->
            return Pair(candidate, oid)
        }
    }

    throw MuonGitException.NotFound("could not determine remote default branch")
}

private fun defaultPushRefspecs(gitDir: File): List<String> {
    val head = readReference(gitDir, "HEAD")
    if (!head.startsWith("ref: ")) {
        throw MuonGitException.InvalidSpec("HEAD is detached; provide push refspecs")
    }
    val target = head.removePrefix("ref: ").trim()
    return listOf("$target:$target")
}

private fun isFastForward(gitDir: File, oldOID: OID, newOID: OID): Boolean {
    if (oldOID.isZero) {
        return true
    }
    return mergeBase(gitDir, oldOID, newOID) == oldOID
}

private fun updateTrackingRefsAfterPush(
    gitDir: File,
    remoteName: String,
    updates: List<PushUpdate>,
): Int {
    var updated = 0
    for (update in updates) {
        if (!update.dstRef.startsWith("refs/heads/")) {
            continue
        }
        val branch = update.dstRef.removePrefix("refs/heads/")
        writeReference(gitDir, "refs/remotes/$remoteName/$branch", update.srcOid)
        updated++
    }
    return updated
}

private fun startsWithBytes(bytes: ByteArray, prefix: ByteArray): Boolean {
    if (bytes.size < prefix.size) {
        return false
    }
    for (index in prefix.indices) {
        if (bytes[index] != prefix[index]) {
            return false
        }
    }
    return true
}

File: swift/src/Fetch.swift
// Fetch.swift - Fetch, push, and clone operations
// Parity: libgit2 src/libgit2/fetch.c, push.c, clone.c

import Foundation

// MARK: - Fetch

/// Result of computing fetch wants.
public struct FetchNegotiation {
    /// OIDs we need to fetch.
    public let wants: [OID]
    /// OIDs we already have.
    public let haves: [OID]
    /// Remote refs that matched the fetch refspecs.
    public let matchedRefs: [MatchedRef]
}

/// A remote ref matched against a fetch refspec.
public struct MatchedRef {
    public let remoteName: String
    public let localName: String
    public let oid: OID
}

/// Match a ref name against a refspec pattern (supports trailing glob).
func refspecMatch(_ name: String, pattern: String) -> String? {
    if pattern.hasSuffix("*") {
        let prefix = String(pattern.dropLast())
        if name.hasPrefix(prefix) {
            return String(name.dropFirst(prefix.count))
        }
        return nil
    } else if name == pattern {
        return ""
    } else {
        return nil
    }
}

/// Apply a refspec to map a remote ref name to a local ref name.
public func applyRefspec(_ remoteName: String, refspec: String) -> String? {
    guard let (_, src, dst) = parseRefspec(refspec) else { return nil }
    guard let matched = refspecMatch(remoteName, pattern: src) else { return nil }

    if let dstPrefix = dst.hasSuffix("*") ? String(dst.dropLast()) : nil {
        return "\(dstPrefix)\(matched)"
    } else {
        return dst
    }
}

/// Compute which objects we need to fetch from the remote.
public func computeFetchWants(
    remoteRefs: [RemoteRef],
    refspecs: [String],
    gitDir: String
) -> FetchNegotiation {
    var wants: [OID] = []
    var matchedRefs: [MatchedRef] = []
    var seen = Set<String>()

    for rref in remoteRefs {
        for refspec in refspecs {
            if let localName = applyRefspec(rref.name, refspec: refspec) {
                matchedRefs.append(MatchedRef(
                    remoteName: rref.name,
                    localName: localName,
                    oid: rref.oid
                ))

                let alreadyHave: Bool
                if let localOid = try? resolveReference(gitDir: gitDir, name: localName) {
                    alreadyHave = localOid == rref.oid
                } else {
                    alreadyHave = false
                }

                if !alreadyHave && !seen.contains(rref.oid.hex) {
                    seen.insert(rref.oid.hex)
                    wants.append(rref.oid)
                }
            }
        }
    }

    let haves = collectLocalRefs(gitDir: gitDir)

    return FetchNegotiation(wants: wants, haves: haves, matchedRefs: matchedRefs)
}

/// Collect all local ref OIDs for negotiation.
func collectLocalRefs(gitDir: String) -> [OID] {
    var oids: [OID] = []
    let fm = FileManager.default
    for dir in ["refs/heads", "refs/remotes"] {
        let refDir = (gitDir as NSString).appendingPathComponent(dir)
        collectRefsRecursive(refDir, oids: &oids, fm: fm)
    }
    return oids
}

private func collectRefsRecursive(_ dir: String, oids: inout [OID], fm: FileManager) {
    guard let items = try? fm.contentsOfDirectory(atPath: dir) else { return }
    for item in items {
        let path = (dir as NSString).appendingPathComponent(item)
        var isDir: ObjCBool = false
        fm.fileExists(atPath: path, isDirectory: &isDir)
        if isDir.boolValue {
            collectRefsRecursive(path, oids: &oids, fm: fm)
        } else if let content = try? String(contentsOfFile: path, encoding: .utf8) {
            let hex = content.trimmingCharacters(in: .whitespacesAndNewlines)
            if hex.count == 40 {
                oids.append(OID(hex: hex))
            }
        }
    }
}

/// Update local refs after a successful fetch.
public func updateRefsFromFetch(gitDir: String, matchedRefs: [MatchedRef]) throws -> Int {
    var updated = 0
    for mref in matchedRefs {
        try writeReference(gitDir: gitDir, name: mref.localName, oid: mref.oid)
        updated += 1
    }
    return updated
}

// MARK: - Push

/// A ref update for push.
public struct PushUpdate: Equatable {
    public let srcRef: String
    public let dstRef: String
    public let srcOid: OID
    public let dstOid: OID
    public let force: Bool
}

/// Compute push updates.
public func computePushUpdates(
    pushRefspecs: [String],
    gitDir: String,
    remoteRefs: [RemoteRef]
) throws -> [PushUpdate] {
    var updates: [PushUpdate] = []

    for refspec in pushRefspecs {
        guard let (force, src, dst) = parseRefspec(refspec) else {
            throw MuonGitError.invalidObject("invalid push refspec: \(refspec)")
        }

        let srcOid = try resolveReference(gitDir: gitDir, name: src)
        let dstOid = remoteRefs.first(where: { $0.name == dst })?.oid ?? OID.zero

        updates.append(PushUpdate(
            srcRef: src,
            dstRef: dst,
            srcOid: srcOid,
            dstOid: dstOid,
            force: force
        ))
    }

    return updates
}

/// Build a push report string.
public func buildPushReport(_ updates: [PushUpdate]) -> String {
    var report = ""
    for u in updates {
        report += "\(u.dstOid.hex) \(u.srcOid.hex) \(u.dstRef)\n"
    }
    return report
}

public struct FetchOptions {
    public let refspecs: [String]?
    public let transport: TransportOptions

    public init(refspecs: [String]? = nil, transport: TransportOptions = TransportOptions()) {
        self.refspecs = refspecs
        self.transport = transport
    }
}

public struct FetchResult {
    public let advertisedRefs: [RemoteRef]
    public let capabilities: ServerCapabilities
    public let matchedRefs: [MatchedRef]
    public let updatedRefs: Int
    public let indexedPack: IndexedPack?
}

public struct PushOptions {
    public let refspecs: [String]?
    public let transport: TransportOptions

    public init(refspecs: [String]? = nil, transport: TransportOptions = TransportOptions()) {
        self.refspecs = refspecs
        self.transport = transport
    }
}

public struct PushResult {
    public let advertisedRefs: [RemoteRef]
    public let updatedTrackingRefs: Int
    public let report: String
}

// MARK: - Clone

/// Options for clone.
public struct CloneOptions {
    public let remoteName: String
    public let branch: String?
    public let bare: Bool
    public let transport: TransportOptions

    public init(
        remoteName: String = "origin",
        branch: String? = nil,
        bare: Bool = false,
        transport: TransportOptions = TransportOptions()
    ) {
        self.remoteName = remoteName
        self.branch = branch
        self.bare = bare
        self.transport = transport
    }
}

/// Set up a new repository for clone: init repo, add remote, configure HEAD.
public func cloneSetup(path: String, url: String, options: CloneOptions = CloneOptions()) throws -> Repository {
    let repo = try Repository.create(at: path, bare: options.bare)

    _ = try addRemote(gitDir: repo.gitDir, name: options.remoteName, url: url)

    if let branch = options.branch {
        let target = "refs/heads/\(branch)"
        try writeSymbolicReference(gitDir: repo.gitDir, name: "HEAD", target: target)
    }

    return repo
}

/// After fetching, set up HEAD and the default branch for a clone.
public func cloneFinish(gitDir: String, remoteName: String, defaultBranch: String, headOid: OID) throws {
    let localBranch = "refs/heads/\(defaultBranch)"
    let remoteRef = "refs/remotes/\(remoteName)/\(defaultBranch)"

    try writeReference(gitDir: gitDir, name: localBranch, oid: headOid)
    try writeReference(gitDir: gitDir, name: remoteRef, oid: headOid)
    try writeSymbolicReference(gitDir: gitDir, name: "HEAD", target: localBranch)
}

/// Extract the default branch from server capabilities.
public func defaultBranchFromCaps(_ caps: ServerCapabilities) -> String? {
    guard let symref = caps.get("symref") else { return nil }
    let parts = symref.split(separator: ":", maxSplits: 1).map(String.init)
    guard parts.count == 2, parts[0] == "HEAD" else { return nil }
    let target = parts[1]
    if target.hasPrefix("refs/heads/") {
        return String(target.dropFirst("refs/heads/".count))
    }
    return nil
}

public func cloneRepository(from url: String, to path: String, options: CloneOptions = CloneOptions()) throws -> Repository {
    let repo = try cloneSetup(path: path, url: url, options: options)
    let fetch = try fetchRemote(repository: repo, remoteName: options.remoteName, options: FetchOptions(transport: options.transport))
    let (branch, headOID) = try resolveCloneHead(fetch, branch: options.branch)
    try cloneFinish(gitDir: repo.gitDir, remoteName: options.remoteName, defaultBranch: branch, headOid: headOID)
    if !options.bare {
        _ = try reset(gitDir: repo.gitDir, workdir: repo.workdir, spec: "HEAD", mode: .hard)
    }
    return repo
}

public func fetchRemote(
    repository: Repository,
    remoteName: String,
    options: FetchOptions = FetchOptions()
) throws -> FetchResult {
    let remote = try getRemote(gitDir: repository.gitDir, name: remoteName)
    let advertisement = try advertiseUploadPack(url: remote.url, options: options.transport)
    let refspecs = options.refspecs ?? remote.fetchRefspecs
    let negotiation = computeFetchWants(remoteRefs: advertisement.refs, refspecs: refspecs, gitDir: repository.gitDir)

    if negotiation.wants.isEmpty {
        let updated = try updateRefsFromFetch(gitDir: repository.gitDir, matchedRefs: negotiation.matchedRefs)
        return FetchResult(
            advertisedRefs: advertisement.refs,
            capabilities: advertisement.capabilities,
            matchedRefs: negotiation.matchedRefs,
            updatedRefs: updated,
            indexedPack: nil
        )
    }

    let request = buildWantHave(
        wants: negotiation.wants,
        haves: negotiation.haves,
        caps: fetchCapabilities(advertisement.capabilities)
    )
    let response = try uploadPack(url: remote.url, request: request, options: options.transport)
    let indexedPack = try extractPackFromFetchResponse(response)
        .map { try indexPackToODB(gitDir: repository.gitDir, packBytes: Array($0)) }
    let updated = try updateRefsFromFetch(gitDir: repository.gitDir, matchedRefs: negotiation.matchedRefs)

    return FetchResult(
        advertisedRefs: advertisement.refs,
        capabilities: advertisement.capabilities,
        matchedRefs: negotiation.matchedRefs,
        updatedRefs: updated,
        indexedPack: indexedPack
    )
}

public func pushRemote(
    repository: Repository,
    remoteName: String,
    options: PushOptions = PushOptions()
) throws -> PushResult {
    let remote = try getRemote(gitDir: repository.gitDir, name: remoteName)
    let advertisement = try advertiseReceivePack(url: remote.url, options: options.transport)
    let refspecs = try options.refspecs ?? defaultPushRefspecs(gitDir: repository.gitDir)
    let updates = try computePushUpdates(pushRefspecs: refspecs, gitDir: repository.gitDir, remoteRefs: advertisement.refs)

    for update in updates {
        if update.force {
            continue
        }
        if !(try isFastForward(gitDir: repository.gitDir, oldOID: update.dstOid, newOID: update.srcOid)) {
            throw MuonGitError.notFastForward
        }
    }

    let pack = try buildPackFromOIDs(
        gitDir: repository.gitDir,
        roots: updates.map { $0.srcOid },
        exclude: advertisement.refs.map { $0.oid }
    )
    let request = buildPushRequest(updates: updates, pack: pack, capabilities: advertisement.capabilities)
    let response = try receivePack(url: remote.url, request: Data(request), options: options.transport)
    let report = try parsePushResponse(response)
    let updatedTrackingRefs = try updateTrackingRefsAfterPush(gitDir: repository.gitDir, remoteName: remoteName, updates: updates)

    return PushResult(
        advertisedRefs: advertisement.refs,
        updatedTrackingRefs: updatedTrackingRefs,
        report: report
    )
}

public extension Repository {
    func fetch(remoteName: String, options: FetchOptions = FetchOptions()) throws -> FetchResult {
        try fetchRemote(repository: self, remoteName: remoteName, options: options)
    }

    func push(remoteName: String, options: PushOptions = PushOptions()) throws -> PushResult {
        try pushRemote(repository: self, remoteName: remoteName, options: options)
    }
}

private func fetchCapabilities(_ caps: ServerCapabilities) -> [String] {
    var requested: [String] = []
    if caps.has("side-band-64k") {
        requested.append("side-band-64k")
    } else if caps.has("side-band") {
        requested.append("side-band")
    }
    if caps.has("ofs-delta") {
        requested.append("ofs-delta")
    }
    if caps.has("include-tag") {
        requested.append("include-tag")
    }
    return requested
}

private func extractPackFromFetchResponse(_ response: Data) throws -> Data? {
    let (lines, consumed) = try unwrap(pktLineDecode(response))
    var pack = Data()

    for line in lines {
        guard case let .data(data) = line else { continue }
        if data.starts(with: Data("ACK ".utf8)) || data == Data("NAK\n".utf8) {
            continue
        }
        guard let first = data.first else { continue }
        switch first {
        case 1:
            pack.append(data.dropFirst())
        case 2:
            break
        case 3:
            throw MuonGitError.invalid(String(decoding: data.dropFirst(), as: UTF8.self).trimmingCharacters(in: .whitespacesAndNewlines))
        default:
            break
        }
    }

    if !pack.isEmpty {
        return pack
    }

    if consumed < response.count {
        let trailing = response.subdata(in: consumed..<response.count)
        if trailing.starts(with: Data("PACK".utf8)) {
            return trailing
        }
    }

    return nil
}

private func buildPushRequest(updates: [PushUpdate], pack: [UInt8], capabilities: ServerCapabilities) -> [UInt8] {
    var requestedCaps = ["report-status"]
    if capabilities.has("ofs-delta") {
        requestedCaps.append("ofs-delta")
    }

    var out = Data()
    for (index, update) in updates.enumerated() {
        let line: String
        if index == 0 {
            line = "\(update.dstOid.hex) \(update.srcOid.hex) \(update.dstRef)\u{0}\(requestedCaps.joined(separator: " "))\n"
        } else {
            line = "\(update.dstOid.hex) \(update.srcOid.hex) \(update.dstRef)\n"
        }
        out.append(pktLineEncode(Data(line.utf8)))
    }
    out.append(pktLineFlush())
    out.append(Data(pack))
    return Array(out)
}

private func parsePushResponse(_ response: Data) throws -> String {
    let (lines, consumed) = try unwrap(pktLineDecode(response))
    var text = ""

    for line in lines {
        guard case let .data(data) = line else { continue }
        let payload: Data
        if let first = data.first, first == 1 || first == 2 {
            payload = data.dropFirst()
        } else if let first = data.first, first == 3 {
            throw MuonGitError.invalid(String(decoding: data.dropFirst(), as: UTF8.self).trimmingCharacters(in: .whitespacesAndNewlines))
        } else {
            payload = data
        }
        text += String(decoding: payload, as: UTF8.self)
    }

    if consumed < response.count {
        text += String(decoding: response.subdata(in: consumed..<response.count), as: UTF8.self)
    }

    for line in text.split(separator: "\n").map(String.init) {
        if line.hasPrefix("unpack "), line != "unpack ok" {
            throw MuonGitError.invalid(line)
        }
        if line.hasPrefix("ng ") {
            throw MuonGitError.invalid(String(line.dropFirst(3)))
        }
    }

    return text
}

private func resolveCloneHead(_ fetch: FetchResult, branch: String?) throws -> (String, OID) {
    if let branch {
        let refName = "refs/heads/\(branch)"
        guard let oid = fetch.advertisedRefs.first(where: { $0.name == refName })?.oid else {
            throw MuonGitError.notFound("remote branch '\(branch)' not found")
        }
        return (branch, oid)
    }

    if let branch = defaultBranchFromCaps(fetch.capabilities) {
        let refName = "refs/heads/\(branch)"
        if let oid = fetch.advertisedRefs.first(where: { $0.name == refName })?.oid {
            return (branch, oid)
        }
    }

    if let headOID = fetch.advertisedRefs.first(where: { $0.name == "HEAD" })?.oid,
       let branchRef = fetch.advertisedRefs.first(where: { $0.name.hasPrefix("refs/heads/") && $0.oid == headOID }) {
        return (String(branchRef.name.dropFirst("refs/heads/".count)), headOID)
    }

    for candidate in ["main", "master"] {
        let refName = "refs/heads/\(candidate)"
        if let oid = fetch.advertisedRefs.first(where: { $0.name == refName })?.oid {
            return (candidate, oid)
        }
    }

    throw MuonGitError.notFound("could not determine remote default branch")
}

private func defaultPushRefspecs(gitDir: String) throws -> [String] {
    let head = try readReference(gitDir: gitDir, name: "HEAD")
    guard head.hasPrefix("ref: ") else {
        throw MuonGitError.invalidSpec("HEAD is detached; provide push refspecs")
    }
    let target = String(head.dropFirst(5)).trimmingCharacters(in: .whitespacesAndNewlines)
    return ["\(target):\(target)"]
}

private func isFastForward(gitDir: String, oldOID: OID, newOID: OID) throws -> Bool {
    if oldOID == .zero {
        return true
    }
    return try mergeBase(gitDir: gitDir, oid1: oldOID, oid2: newOID) == oldOID
}

private func updateTrackingRefsAfterPush(gitDir: String, remoteName: String, updates: [PushUpdate]) throws -> Int {
    var updated = 0
    for update in updates where update.dstRef.hasPrefix("refs/heads/") {
        let branch = update.dstRef.dropFirst("refs/heads/".count)
        try writeReference(gitDir: gitDir, name: "refs/remotes/\(remoteName)/\(branch)", oid: update.srcOid)
        updated += 1
    }
    return updated
}

private func unwrap<T>(_ result: Result<T, MuonGitError>) throws -> T {
    switch result {
    case let .success(value): return value
    case let .failure(error): throw error
    }
}
